Nomad v2.0.1 fails on all allocations with acls disabled

### Nomad version
[root@apps1 jonas]$ nomad version
Nomad v2.0.1
BuildDate 2026-05-12T13:14:32Z
Revision 5c11f948ba7e9e709abf7f2ab7895ba5b44feb39

### Operating system and Environment details
Linux, Debian 13

### Issue
All servers and clients keep logging the same error

```
May 13 15:03:59 apps1 nomad[9255]:     2026-05-13T15:03:59.887Z [ERROR] client: error querying updated allocations: error="rpc error: Permission denied"
May 13 15:03:59 apps1 nomad[9255]: client.rpc: error performing RPC to server: error="rpc error: Permission denied" rpc=Alloc.GetAllocs server=10.0.1.6:4647
May 13 15:03:59 apps1 nomad[9255]: client.rpc: error performing RPC to server which is not safe to automatically retry: error="rpc error: Permission denied" rpc=Alloc.GetAllocs server=10.0.1.6:4647
May 13 15:03:59 apps1 nomad[9255]: client: error querying updated allocations: error="rpc error: Permission denied"
```

### Reproduction steps
As far as I can tell, any Nomad cluster I spin up with my configuration is affected. Even a cluster I build from scratch immediately shows this error when I use v2.0.1 while v1.11.3 works fine.

```hcl
[root@apps1 jonas]$ cat /etc/nomad/agent.hcl 

datacenter = "RZ19"
enable_syslog = true

client {
    enabled = true

    # pin the "network device fingerprinting" to the network device connected
    # to the internal network. This is necessary so nomad doesn't instruct
    # docker to just bind any container to eth0's first IP.
    #
    network_interface = "enp7s0"
    preferred_address_family = "ipv4"
    
    # this allows access to the cluster's root CA on every host (for .local services, for example)
    host_volume "host-ca-bundle" {
        path      = "/etc/ssl/certs/ca-certificates.crt"
        read_only = true
    }

    cni_path = "/usr/local/lib/nomad/"
    cni_config_dir = "/etc/nomad/cni/"
}

consul {
    address = "127.0.0.1:8500"
    token = "***"
}
```

```hcl
[root@apps1 jonas]$ cat /etc/nomad/server.hcl 

server {
    enabled = true
    bootstrap_expect = 3
}

acl {
    enabled = false
}
```

```hcl
[root@apps1 jonas]$ cat /etc/nomad/common.hcl 
bind_addr = "10.0.1.4"
data_dir  = "/var/lib/nomad"

disable_update_check = true


advertise {
    # Defaults to the node's hostname. If the hostname resolves to a loopback
    # address you must manually configure advertise addresses.
    http = "10.0.1.4"
    rpc  = "10.0.1.4"
    serf = "10.0.1.4"
}

plugin "docker" {
    config {
        endpoint = "unix:///var/run/docker.sock"

        extra_labels = ["job_name", "job_id", "task_group_name", "task_name", "namespace", "node_name", "node_id"]

        gc {
            image       = true
            image_delay = "3m"
            container   = true

            dangling_containers {
                enabled        = true
                dry_run        = false
                period         = "5m"
                creation_grace = "5m"
            }
        }

        volumes {
            enabled      = true
            selinuxlabel = "z"
        }

        # allow_privileged is required to run Hetzner CSI drivers, which require privileges to attach volumes.
        # the cap list is the default + net_raw as that's also required by the Hetzner CSI drivers.
        allow_privileged = true
        allow_caps       = [
            "audit_write", "chown", "dac_override", "fowner", "fsetid", "kill", "mknod",
            "net_bind_service", "net_raw", "setfcap", "setgid", "setpcap", "setuid", "sys_chroot"
        ]
    }
}
```

```ini
[root@apps1 jonas]$ systemctl cat nomad
# /etc/systemd/system/nomad.service
[Unit]
Description=Nomad Cluster
Wants=basic.target
After=basic.target network.target
# StartLimitInterval: try to restart for unlimited times, because this is an essential service!
StartLimitIntervalSec=0

[Service]
# User=nomad when we can switch to a dedicated nomad user
User=root
# Group=nomad when we can switch to a dedicated nomad group
Group=root
ExecStart=/usr/local/bin/nomad agent -config=/etc/nomad/common.hcl -config=/etc/nomad/server.hcl -config=/etc/nomad/agent.hcl

ExecReload=/bin/kill -HUP $MAINPID
KillMode=process
Restart=on-failure
PIDFile=/run/nomad/nomad.pid
RestartSec=42s

[Install]
WantedBy=multi-user.target
```

#### Expected Result
This exact configuration works fine with Nomad v1.11.3.

#### Actual Result
All allocations fail with the RPC error.

Same issue.
OS: Linux, Ubuntu 22
1 server

Updating from v1.11.3 to v2.0.1. I got a lot of ERROR logs: `[ERROR] client.rpc: error performing RPC to server which is not safe to automatically retry: error="rpc error: Permission denied" rpc=Alloc.GetAllocs server=x.x.x.x`, after server update to v2.0.1. 
Client update from 1.11.3 to 2.0.1 situation didn't changed.

```
May 14 15:16:58 client1 nomad[6836]:     2026-05-14T15:16:58.869Z [ERROR] client.rpc: error performing RPC to server which is not safe to automatically retry: error="rpc error: Permission denied" rpc=Alloc.GetAllocs server=192.168.1.100:4647
```
Ok so it looks like I've been able to replicate the issue, thanks for the job files @jdelic . This issue happens when using `node_pool = "all"`. And the change is related to a downscoping of client privileges to only be allowed to make RPC's for allocations within that nodes node_pool, [here](https://github.com/hashicorp/nomad/pull/27718). I'll try to get a fix up here soon.

Sorry about all the back and forth on reproducing this.. And I really appreciate the help from you both.

--- a/acl/acl.go
+++ b/acl/acl.go
@@ -1017,7 +1017,8 @@ func (a *ACL) AllowClientOp(pool string) bool {
 	if a.client == PolicyDeny {
 		return false
 	}
-	return pool == a.pool
+	// hardcode "all" until we can use node.IsInPool
+	return pool == a.pool || pool == "all"
 }
 
 // IsManagement checks if this represents a management token

--- a/acl/acl.go
+++ b/acl/acl.go
@@ -1017,7 +1017,6 @@ func (a *ACL) AllowClientOp(pool string) bool {
 	if a.client == PolicyDeny {
 		return false
 	}
-	// hardcode "all" until we can use node.IsInPool
 	return pool == a.pool || pool == "all"
 }
 

--- a/acl/acl_test.go
+++ b/acl/acl_test.go
@@ -103,6 +103,14 @@ func TestACLManagement(t *testing.T) {
 	must.False(t, acl.AllowClientOp("my-pool"))
 }
 
+func TestClientACL(t *testing.T) {
+	acl := NewClientACL("test-pool")
+
+	must.True(t, acl.AllowClientOp("test-pool"))
+	must.True(t, acl.AllowClientOp("all")) // tests node_pool = "all"
+	must.False(t, acl.AllowClientOp("test-not-my-pool"))
+}
+
 func TestACLMerge(t *testing.T) {
 	ci.Parallel(t)
 

--- a/acl/acl_test.go
+++ b/acl/acl_test.go
@@ -103,14 +103,6 @@ func TestACLManagement(t *testing.T) {
 	must.False(t, acl.AllowClientOp("my-pool"))
 }
 
-func TestClientACL(t *testing.T) {
-	acl := NewClientACL("test-pool")
-
-	must.True(t, acl.AllowClientOp("test-pool"))
-	must.True(t, acl.AllowClientOp("all")) // tests node_pool = "all"
-	must.False(t, acl.AllowClientOp("test-not-my-pool"))
-}
-
 func TestACLMerge(t *testing.T) {
 	ci.Parallel(t)
 
@@ -229,6 +221,14 @@ func TestACLMerge(t *testing.T) {
 
 func TestACLAllowClientOp_NodePoolScoped(t *testing.T) {
 	ci.Parallel(t)
+
+	t.Run("client acl works with node pool 'all'", func(t *testing.T) {
+		acl := NewClientACL("test-pool")
+
+		must.True(t, acl.AllowClientOp("test-pool"))
+		must.True(t, acl.AllowClientOp("all")) // tests node_pool = "all"
+		must.False(t, acl.AllowClientOp("test-not-my-pool"))
+	})
 
 	t.Run("management", func(t *testing.T) {
 		acl, err := NewACL(true, nil)

--- a/e2e/scheduler_system/input/system_job_all_pools.nomad
+++ b/e2e/scheduler_system/input/system_job_all_pools.nomad
@@ -1,0 +1,40 @@
+# Copyright IBM Corp. 2015, 2025
+# SPDX-License-Identifier: BUSL-1.1
+
+job "system_job" {
+  datacenters = ["dc1", "dc2"]
+
+  type = "system"
+
+  # Many system jobs are run in every node pool
+  # lets test this job runs as expected
+  node_pool = "all"
+
+  constraint {
+    attribute = "${attr.kernel.name}"
+    value     = "linux"
+  }
+
+  group "system_job_group" {
+    count = 1
+
+    restart {
+      attempts = 10
+      interval = "1m"
+
+      delay = "2s"
+      mode  = "delay"
+    }
+
+    task "system_task" {
+      driver = "docker"
+
+      config {
+        image = "busybox:1"
+
+        command = "/bin/sh"
+        args    = ["-c", "sleep 15000"]
+      }
+    }
+  }
+}

--- a/e2e/scheduler_system/input/system_job_all_pools.nomad
+++ b/e2e/scheduler_system/input/system_job_all_pools.nomad
@@ -1,4 +1,4 @@
-# Copyright IBM Corp. 2015, 2025
+# Copyright IBM Corp. 2015, 2026
 # SPDX-License-Identifier: BUSL-1.1
 
 job "system_job" {

--- a/e2e/scheduler_system/systemsched_test.go
+++ b/e2e/scheduler_system/systemsched_test.go
@@ -1,4 +1,4 @@
-// Copyright IBM Corp. 2015, 2025
+// Copyright IBM Corp. 2015, 2026
 // SPDX-License-Identifier: BUSL-1.1
 
 package scheduler_system

--- a/e2e/scheduler_system/systemsched_test.go
+++ b/e2e/scheduler_system/systemsched_test.go
@@ -21,9 +21,39 @@ func TestSystemScheduler(t *testing.T) {
 		cluster3.LinuxClients(3),
 	)
 
+	t.Run("testSystemJobAllNodePools", testSystemJobAllNodePools)
 	t.Run("testJobUpdateOnIneligibleNode", testJobUpdateOnIneligbleNode)
 	t.Run("testCanaryUpdate", testCanaryUpdate)
 	t.Run("testCanaryDeploymentToAllEligibleNodes", testCanaryDeploymentToAllEligibleNodes)
+}
+
+func testSystemJobAllNodePools(t *testing.T) {
+	job, cleanup := jobs3.Submit(t,
+		"./input/system_job_all_pools.nomad",
+		jobs3.DisableRandomJobID(),
+		jobs3.Timeout(60*time.Second),
+	)
+	t.Cleanup(cleanup)
+
+	deploymentsApi := job.DeploymentsApi()
+	deploymentsList, _, err := deploymentsApi.List(nil)
+	must.NoError(t, err)
+
+	var deployment *api.Deployment
+	for _, d := range deploymentsList {
+		if d.JobID == job.JobID() && d.Status == api.DeploymentStatusRunning {
+			deployment = d
+		}
+	}
+	must.NotNil(t, deployment)
+
+	// wait for the deployment to become healthy
+	timeout, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	job.WaitForDeploymentFunc(timeout, deployment.ID, func(d *api.Deployment) bool {
+		return d.Status == api.DeploymentStatusSuccessful
+	})
 }
 
 func testJobUpdateOnIneligbleNode(t *testing.T) {

--- a/e2e/scheduler_system/systemsched_test.go
+++ b/e2e/scheduler_system/systemsched_test.go
@@ -48,7 +48,7 @@ func testSystemJobAllNodePools(t *testing.T) {
 	must.NotNil(t, deployment)
 
 	// wait for the deployment to become healthy
-	timeout, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	timeout, cancel := context.WithTimeout(t.Context(), 60*time.Second)
 	defer cancel()
 
 	job.WaitForDeploymentFunc(timeout, deployment.ID, func(d *api.Deployment) bool {

